src/main/java/bio/terra/common/ValidationUtils.java

```diff
@@ -80,6 +81,148 @@ public static String requireNotBlank(String value, String errorMsg) {
     return value;
   }
 
+  private static boolean compatibleDataType(
+      TableDataType fromDataType,
+      TableDataType toDataType,
+      CloudPlatformWrapper cloudPlatformWrapper) {
+    if (cloudPlatformWrapper.isGcp()) {
```

Comment:
A few comments about this code.

- the code is, generally, switching over `fromDataType` so it can call `xxx.contains(toDataType)`. changing the code to call `contains(toDataType)` once, at `return` would make the code intent clearer, IMO
- changing the code to use a switch expression so that each switch generates a list would ensure that the list of switch labels are exhaustive
- once you have an exhaustive switch, you can remove `default` and specify every case explicitly. This is better than using `default`, since whenever the enum `TableDataType ` is changed, the switch will no longer compile until the new case is handled.

One way to write this is
```
  private static boolean isCompatibleDataType(
      TableDataType fromDataType,
      TableDataType toDataType,
      CloudPlatformWrapper cloudPlatformWrapper) {
    final List<TableDataType> compatibleTypes;
    if (cloudPlatformWrapper.isGcp()) {
      // https://cloud.google.com/bigquery/docs/reference/standard-sql/conversion_rules
      compatibleTypes = switch (fromDataType) {
        case DATE, DATETIME -> List.of(TableDataType.DATE, TableDataType.DATETIME);
        case DIRREF, FILEREF -> List.of(TableDataType.DIRREF, TableDataType.FILEREF);
        case FLOAT, FLOAT64, INTEGER, INT64, NUMERIC -> List.of(
            TableDataType.FLOAT,
            TableDataType.FLOAT64,
            TableDataType.INTEGER,
            TableDataType.INT64,
            TableDataType.NUMERIC);
        case STRING, TEXT -> List.of(TableDataType.STRING, TableDataType.TEXT);
        case RECORD, BOOLEAN, BYTES, TIME, TIMESTAMP -> List.of(fromDataType);
      };
    } else if (cloudPlatformWrapper.isAzure()) {
      // Following rules for implicit conversion as defined here:
      // https://learn.microsoft.com/en-us/sql/t-sql/data-types/data-type-conversion-database-engine?view=sql-server-ver16
      compatibleTypes = switch (fromDataType) {
        case BOOLEAN, FLOAT, FLOAT64, INTEGER, INT64, NUMERIC -> List.of(
            TableDataType.BOOLEAN,
            TableDataType.BYTES,
            TableDataType.FLOAT,
            TableDataType.FLOAT64,
            TableDataType.INTEGER,
            TableDataType.INT64,
            TableDataType.NUMERIC,
            TableDataType.TEXT,
            TableDataType.STRING);
        case BYTES -> List.of(
            TableDataType.BOOLEAN,
            TableDataType.BYTES,
            TableDataType.INTEGER,
            TableDataType.INT64,
            TableDataType.TEXT,
            TableDataType.STRING);
        case DATE -> List.of(
            TableDataType.DATE,
            TableDataType.DATETIME,
            TableDataType.TIMESTAMP,
            TableDataType.STRING,
            TableDataType.TEXT);
        case DATETIME, TIMESTAMP -> List.of(
            TableDataType.DATE,
            TableDataType.DATETIME,
            TableDataType.TIMESTAMP,
            TableDataType.STRING,
            TableDataType.TEXT,
            TableDataType.TIME);
        case DIRREF, FILEREF -> List.of(TableDataType.DIRREF, TableDataType.FILEREF);
        case TEXT, STRING -> List.of(
            TableDataType.BOOLEAN,
            TableDataType.BYTES,
            TableDataType.FLOAT,
            TableDataType.FLOAT64,
            TableDataType.INTEGER,
            TableDataType.INT64,
            TableDataType.NUMERIC,
            TableDataType.TEXT,
            TableDataType.STRING,
            TableDataType.DATE,
            TableDataType.DATETIME,
            TableDataType.TIMESTAMP,
            TableDataType.TIME);
        case TIME -> List.of(
            TableDataType.DATETIME,
            TableDataType.TIMESTAMP,
            TableDataType.STRING,
            TableDataType.TEXT,
            TableDataType.TIME);
        case RECORD -> List.of();
      };
    } else {
      compatibleTypes = List.of();
    }
    return compatibleTypes.contains(toDataType);
  }
```

Reply:
I knew there had to be a better way to do this sort of check, and I think this is much cleaner. Thank you for the detailed and thoughtful solution!